test/metabase/driver/query_processor/expand_resolve_test.clj

```diff
@@ -0,0 +1,286 @@
+(defn- obj->map [o]
+  (if-not (coll? o)
+    o
+    (cond
+      (or (vector? o)
+          (list? o))    (vec (for [v o]
+                               (obj->map v)))
+      (set? o)          (set (for [v o]
+                               (obj->map v)))
+      (map? o)          (into {} (for [[k v] o]
+                                   {k (obj->map v)})))))
+
+
+;; basic rows query w/ filter
+(expect
+  [;; expanded form
+   {:database (id)
+    :type     :query
+    :query    {:source-table (id :venues)
+               :filter       {:filter-type :>
+                              :field {:field-id (id :venues :price)
+                                      :fk-field-id nil
+                                      :datetime-unit nil}
+                              :value {:field-placeholder {:field-id (id :venues :price)
+                                                          :fk-field-id nil
+                                                          :datetime-unit nil}
+                                      :value 1}}}}
+   ;; resolved form
+   {:database {:name "test-data",
+              :details {:short-lived? nil, :db "mem:test-data;USER=GUEST;PASSWORD=guest"},
+              :id (id),
+              :engine :h2},
+   :type :query,
+   :query {:source-table {:schema "PUBLIC", :name "VENUES", :id (id :venues)},
+           :filter {:filter-type :>,
+                    :field {:field-id (id :venues :price),
+                            :field-name "PRICE",
+                            :field-display-name "Price",
+                            :base-type :IntegerField,
+                            :special-type :category,
+                            :visibility-type :normal,
+                            :table-id (id :venues),
+                            :schema-name "PUBLIC",
+                            :table-name "VENUES",
+                            :position nil,
+                            :description nil,
+                            :parent-id nil,
+                            :parent nil},
+                    :value {:value 1,
+                            :field {:field-id (id :venues :price),
+                                    :field-name "PRICE",
+                                    :field-display-name "Price",
+                                    :base-type :IntegerField,
+                                    :special-type :category,
+                                    :visibility-type :normal,
+                                    :table-id (id :venues),
+                                    :schema-name "PUBLIC",
+                                    :table-name "VENUES",
+                                    :position nil,
+                                    :description nil,
+                                    :parent-id nil,
+                                    :parent nil}}},
+           :join-tables nil},
+   :fk-field-ids #{},
+   :table-ids #{(id :venues)}}]
+  (let [expanded-form (expand/expand {:database (id)
```

Comment:
it's preferable to use the QL forms directly in combination with `metabase.test.data/query` since it's more concise and easier to refactor down the road

```clojure
(ql/expand (ql/wrap-inner-query (data/query venues
                                  (ql/filter (ql/and (ql/> $price 1))))))
```